feat: improve bug template

pichlermarc opened the pull request:
## Which problem is this PR solving?

**We often get bug reports that don't give full details needed to reproduce.** 

There's **new wording** in the placeholder **asking for steps to reproduce**. There's now more emphasis on asking for reproducer repositories, as well as explaining why this is helpful. 

I also added some more emphasis on **build instructions and config files for tooling**. It's somewhat common that users are very familiar with their stack and assume that we are too. I added a reminder to let them know that JavaScript is a complex ecosystem and that build instructions are immensely helpful in trying to resolve their issue

It's also quite common that users use unsupported runtimes - so I've introduced an optional field where they have to enter their **runtime and runtime version** so that we can spot it more easily when this is the case.

Finally there's some bugs that only surface on specific Operating Systems - so there's a new optional field for **OS and version**, which should be relatively easy for the user to provide.

## How this looks like

GitHub preview does not fully represent how the issue templates actually looks like, so here's a screenshot of how it will look like (this is on a private repo so asterisks for required fields are missing - they will be there on this repo, since it's public)

![image](https://github.com/user-attachments/assets/b369477d-d54d-4f7a-8076-36766254ae07)

maryliag commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,106 @@ body:
-        providing as much context as you can in order to help us
-        triage and track down your bug as quickly as possible.
-
-        Before filing a bug, please be sure you have searched through
```
I see this part was removed, but I still think is useful to have it

pichlermarc replied:
Agreed :) 
Added back in https://github.com/open-telemetry/opentelemetry-js/pull/5086/commits/9dc41da9e90d234fb8ee5b5d8e5773797dbb2b7c

maryliag commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,106 @@ body:
+  - type: markdown
+    attributes:
+      value: |
+        ## Bug Reproduciton
```
```suggestion
        ## Bug Reproduction
```

maryliag commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,106 @@ body:
+    id: bundlng
+    attributes:
+      description: |
+        Bundling removes `require` and `import` statements, which results in most `@opentelemetry/instrumentation-*` packages not working as they rely on the presense
```
```suggestion
        Bundling removes `require` and `import` statements, which results in most `@opentelemetry/instrumentation-*` packages not working as they rely on the presence
```

maryliag commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,106 @@ body:
+        > We all know the complexities of developing with JavaScript/TypeScript. **Be specific** about your setup and **DO NOT** assume that the
+        > person handling your bug report knows how to use your specific combination of tooling. Always provide build instructions and config files.
       value: |
         ## Steps to Reproduce
```
this part on your image looks a little weird, not sure if intentional

pichlermarc replied:
Hmm, do you mean because it looks smaller + the bold text being gray instead of black? :thinking: 
I think that's happening because it's part of the description for the box instead of it being a top-level "markdown" element. 

I don't think I can change this, because that would mean that I'd have to move it before the "What happened" header, which makes it feel somewhat out-of-place. :disappointed:

maryliag replied:
<img width="740" alt="Screenshot 2024-11-05 at 10 14 16 AM" src="https://github.com/user-attachments/assets/7cc5c5a2-852f-4dd9-8695-f570d8fcf02d">

This part looks weird, because the `## Steps` suppose to look like a title, but is just looks like it was added by mistake. Having markdown elements such as `##` showing instead of being normally rendered up looks weird to me.

Now I'm realizing this is suppose to be the placeholder. Is that it? So people can add their info on it? In that case, maybe some changing in formatting, so removing the `##` and the `<!--`, this way is just a message of what you should add, without markdown elements that won't be rendered

pichlermarc replied:
Hmm, [this is actually how it is today](https://github.com/open-telemetry/opentelemetry-js/issues/new?assignees=&labels=bug%2Ctriage&projects=&template=bug_report.yaml). I was considering splitting it up into multiple textareas, but it ends up looking quite a bit weirder. 🤔 

FWIW it looks better when the user actually clicks into the textarea. It'll become a free-text markdown area with preview like so:

**Before clicking**
![image](https://github.com/user-attachments/assets/ee6b0c9b-dafe-4a68-8917-0ae157b92295)
**After clicking**
![image](https://github.com/user-attachments/assets/07c4cecd-dc51-4159-ab36-97d72d1659f5)

maryliag replied:
Ahh I see! Not too bad then! Thank you for looking into it!

pichlermarc commented:
@maryliag thanks for the feedback. :raised_hands: 
I've addressed some of the comments and also updated the screenshot with the new changes. I'd appreciate you having another look.  :slightly_smiling_face:

maryliag reviewed (approved):
Thanks for making the changes!
I just have this nit about the formatting of the steps that you can change or not, otherwise LGTM

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
+    id: bundling
+    attributes:
+      description: |
+        Bundling removes `require` and `import` statements, which results in most `@opentelemetry/instrumentation-*` packages not working as they rely on the presence
+        of these statements to function.
```
```suggestion
        Bundling removes `require` and `import` statements, which results in most `@opentelemetry/instrumentation-*` packages not working as they rely on the presence of these statements to function.
```

It looked to me in the screenshot that having multiple lines here results in odd line breaks in the rendered output.

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
+        of these statements to function.
+      label: Bundling and `@opentelemetry/instrumentation-*`packages
+      options:
+        - label: I understand that `@opentelemetry/instrumentation-*` packages are not intended to work in bundles apps (with a few exceptions).
```
```suggestion
        - label: I understand that `@opentelemetry/instrumentation-*` packages are not intended to work in bundled apps (with a few exceptions).
```

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
+      label: Bundling and `@opentelemetry/instrumentation-*`packages
+      options:
+        - label: I understand that `@opentelemetry/instrumentation-*` packages are not intended to work in bundles apps (with a few exceptions).
+          required: true
```
I wonder if `required: true` is a little harsh here for a report on something that has nothing to do with bundling.

pichlermarc replied:
We can also leave it out for now, WDYT?

Unfortunately, there's no mechanism that allows us to conditionally show a new checkbox when another was checked. I'd be pretty cool to have a checkbox "I'm using a bundler" which makes another checkbox pop up that says "I understand that..." and a dropdown that asks which bundler they are using (for any problems that are not instrumentation-package related).

JamieDanielson replied:
Agreed that we can remove the checkbox for now so it doesn't impact non-bundler users reporting a bug

pichlermarc replied:
I removed it now, moving that to a troubleshooting guide that I'll link form here in a follow-up PR.

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
+      label: Updating packages
+      options:
+        - label: I have updated all `@opentelemetry/*` packages but the issue persists
+          required: true
```
I hesitate slightly at `required: true` here. If I want to report an issue specifically with an older version then there is no way to move on from this. (This will be more relevant when we have 2.x current versions *and still supported 1.x versions for a period of time.)

Granted that gives the more lazy bug reporter an "out". I'm conflicted. :)

pichlermarc replied:
I think we'd have to update the bug report template when we release 2.0 - for now this still holds and I believe it would free up a lot of our time. Right now a lot of issues are solved by updating, and we don't support older versions unless they are critical vulnerabilities, even there we've only backported it for 2 versions before (and only once as far as I can remember). I feel like we're also not in a position to change our support to anything other than latest (and last major) anytime soon.

I'd follow up with adding an issue that we update this statement in preparation for the SDK 2.0 release, WDYT?

trentm replied:
Sounds good to tweak this after 2.0.

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
+        > [!IMPORTANT]
+        > **Please provide as much detail as you reasonably can.**
+        > We all know the complexities of developing with JavaScript/TypeScript. **Be specific** about your setup and **DO NOT** assume that the
+        > person handling your bug report knows how to use your specific combination of tooling. Always provide build instructions and config files.
```
I cannot tell from the screenshot, but we may want the latter two lines to be on a single line for better reflow of the rendered Markdown for browser width.

pichlermarc replied:
addressed in https://github.com/open-telemetry/opentelemetry-js/pull/5086/commits/d7245428fff8acda4187c264627cd13569d042a1 🙂

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
+    attributes:
+      value: |
+        ## Bug Reproduction
+        > [!IMPORTANT]
+        >  **If you use a third-party package that re-distributes OpenTelemetry, open the bug ticket with that third party unless you can provide steps to reproduce this with pure OpenTelemetry.**
+        >  Digging into third-party distributions of OpenTelemetry is not in scope for this project.
```
This IMPORTANT section could perhaps move to the top section along with the "Please read all parts of this form carefully. ..." paragraph. Whether it is a 3rd-party distro isn't really scoped to a reproduction.

pichlermarc replied:
Sounds good, I moved it to the top in https://github.com/open-telemetry/opentelemetry-js/pull/5086/commits/d7245428fff8acda4187c264627cd13569d042a1

trentm commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,109 @@ body:
     attributes:
       label: OpenTelemetry Setup Code
-      description: Please provide the code you use to set up OpenTelemetry
+      description: |
+        > [!IMPORTANT]
+        > **If you have provided a reproducer repository above, you may omit this step.**
+
+        **Please provide the code you use to set up OpenTelemetry.**
+        If you use `@opentelemetry/auto-instrumentations-node/register`, please state so here
```
```suggestion
      description: |
        Please provide the code you use to set up OpenTelemetry.
        If you use `@opentelemetry/auto-instrumentations-node/register`, please state so here.
        You may omit this step if you ahve provided a reproducer repository.
```

^^ reformats to have the "OpenTelemetry Setup Code" section formatted like the "package.json" section just below. I don't think the "you may omit this step" note here needs to be called out as "IMPORTANT".

pichlermarc replied:
applied in https://github.com/open-telemetry/opentelemetry-js/pull/5086/commits/d7245428fff8acda4187c264627cd13569d042a1

trentm reviewed (approved):
Happy to have this merged as is, but I have some nits, suggestions, and Qs about "required: true" usage below.

pichlermarc commented:
@trentm I addressed some of your comments, two of them are still open - please let me know about your preferred path forward on those. 

I also updated the screenshot in the title again to reflect the current state 🙂

JamieDanielson commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,105 @@ body:
+      description: |
+        **How to find this information:**
+          - check your `package.json`
+            - if you have `"type": "commonjs"` or no `"type"` then you're likely using CommonJS
```
Consider adding a link to Nodejs docs https://nodejs.org/api/modules.html#enabling

pichlermarc replied:
I removed this again - I'll re-introduce in a follow-up PR when I've completed the troubleshooting guide.

JamieDanielson commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,105 @@ body:
+  - type: markdown
+    attributes:
+      value: |
+        ## Basic Troubleshooting
```
As we talked about offline, it might be more effective if we put all of this (useful!) information into a separate dedicated troubleshooting doc, and link to it from the bug report template

JamieDanielson commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,105 @@ body:
         ## Steps to Reproduce
-
+          <!--
+            Please consider providing a link to a reproducer repository. It is extremely difficult to exactly reproduce problems
+            and doing so can take multiple hours for us. By providing a reproducer you indirectly contribute back to the project
+            and ensure we can get working on fixing your bug as soon as possible instead of spending precious time asking
+            questions and checking responses. You know your tooling and setup and we know OpenTelemetry JavaScript internals.
```
```suggestion
            The most effective way to solve a bug is to provide a link to a reproducer repository. It is extremely difficult to exactly reproduce problems
            and doing so can take multiple hours for us. You know your tooling and setup and we know OpenTelemetry JavaScript internals.
```

pichlermarc replied:
yes it does 🧐🎩 
your thanks for the suggestion, accepted 🙂

JamieDanielson commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -5,38 +5,105 @@ body:
 
         ## Additional Details
+         <!--
+           Anything you can think of that will help us reproduce and understad your problem:
```
```suggestion
           Anything you can think of that will help us reproduce and understand your problem:
```

pichlermarc commented:
Re-requesting review because the template has changed quite a bit since you all had a look.
I got rid of the troubleshooting section and will add another PR once this is merged to add a dedicated troubleshooting doc.
